cmd/juju/action/cancel.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2014-2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package action
+
+import (
+	"github.com/juju/cmd"
+	errors "github.com/juju/errors"
+	"github.com/juju/gnuflag"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/cmd/modelcmd"
+	"github.com/juju/juju/cmd/output"
+)
+
+func NewCancelCommand() cmd.Command {
+	return modelcmd.Wrap(&cancelCommand{})
+}
+
+type cancelCommand struct {
+	ActionCommandBase
+	out          cmd.Output
+	requestedIds []string
+}
+
+// Set up the output.
+func (c *cancelCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.ActionCommandBase.SetFlags(f)
+	c.out.AddFlags(f, "yaml", output.DefaultFormatters)
+}
+
+const cancelDoc = `
+Cancel actions matching given IDs or partial ID prefixes.`
+
+func (c *cancelCommand) Info() *cmd.Info {
+	return &cmd.Info{
+		Name:    "cancel-action",
+		Args:    "<<action ID | action ID prefix>...>",
+		Purpose: "Cancel pending actions.",
+		Doc:     cancelDoc,
+	}
+}
+
+func (c *cancelCommand) Init(args []string) error {
+	c.requestedIds = args
+	return nil
+}
+
+func (c *cancelCommand) Run(ctx *cmd.Context) error {
+	api, err := c.NewActionAPIClient()
+	if err != nil {
+		return err
+	}
+	defer api.Close()
+
+	if len(c.requestedIds) == 0 {
+		return errors.Errorf("no actions specified")
+	}
+
+	var actionTags []names.ActionTag
+	for _, requestedId := range c.requestedIds {
+		requestedActionTags, err := getActionTagsByPrefix(api, requestedId)
+		if err != nil {
+			return err
+		}
+		actionTags = append(actionTags, requestedActionTags...)
+	}
+
+	if len(actionTags) < 1 {
+		return errors.Errorf("no actions found matching any specified ID or prefix")
+	}
+
+	entities := []params.Entity{}
+	for _, tag := range actionTags {
+		entities = append(entities, params.Entity{Tag: tag.String()})
+	}
+
+	actions, err := api.Actions(params.Entities{Entities: entities})
+	if err != nil {
+		return err
+	}
+
+	var pendingActions []params.Action
+	for _, action := range actions.Results {
+		if action.Status == params.ActionPending {
+			pendingActions = append(pendingActions, *action.Action)
+		}
+	}
+
+	if len(pendingActions) < 1 {
+		return errors.Errorf("no pending actions found matching any of specified ID or prefix")
+	}
+
+	actions, err = api.Cancel(params.Actions{Actions: pendingActions})
```

Comment:
What happens if pendingActions contains non-pending actions by the time we get to here? I think it would be preferable to just return an error code that indicates the action is pending/non-cancellable, and have the client ignore those. That way we avoid races, and as a side benefit this version of the client will be able to cancel non-pending actions when we fix Juju to be able to do that.